system/xmllib/libxml_xpathprocessor.cpp

```diff
@@ -775,6 +789,352 @@ extern ICompiledXpath* compileXpath(const char * xpath)
+        return nullptr;
+    return xmlNewNs(node, (const xmlChar *)uri, (const xmlChar *)attname);
+}
+void addChildFromPtree(xmlNodePtr parent, IPropertyTree &tree, const char *name)
+{
+    if (isEmptyString(name))
+        name = tree.queryName();
+    if (*name==':') //invalid xml but maybe valid ptree?
+        name++;
+    StringAttr prefix;
+    const char *colon = strchr(name, ':');
+    if (colon)
+    {
+        prefix.set(name, colon-name);
+        name = colon+1;
+    }
+    xmlNodePtr node = xmlNewChild(parent, nullptr, (const xmlChar *)name, (const xmlChar *)tree.queryProp(nullptr));
+    if (!node)
+        return;
+    xmlNsPtr ns = nullptr;
+    if (tree.getAttributeCount())
+    {
+        Owned<IAttributeIterator> atts = tree.getAttributes();
+        ForEach(*atts)
+        {
+            const char *attname = atts->queryName()+1;
+            if (strncmp(attname, "xmlns", 5)==0)
+            {
+                attname+=5;
+                if (*attname==':')
+                    attname++;
+             xmlNewNs(node, (const xmlChar *)atts->queryValue(), (const xmlChar *)attname);
+            }
+            else
+                xmlNewProp(node, (const xmlChar *)attname, (const xmlChar *)atts->queryValue());
+        }
+    }
+    if (prefix.isEmpty())
+        node->ns = nullptr;
+    else
+    {
+        xmlNsPtr ns = xmlSearchNs(node->doc, node, (const xmlChar *)prefix.str());
+        if (!ns) //invalid but let's repair it
+            ns = xmlNewNs(node, (const xmlChar *)"urn:hpcc:unknown", (const xmlChar *)prefix.str());
+        node->ns = ns;
+    }
+    Owned<IPropertyTreeIterator> children = tree.getElements("*");
+    ForEach(*children)
+        addChildFromPtree(node, children->query(), nullptr);
+}
+
+IPropertyTree *createPtreeFromXmlNode(xmlNodePtr node);
+
+void copyXmlNode(IPropertyTree *tree, xmlNodePtr node)
+{
+    for(xmlAttrPtr att=node->properties;att!=nullptr; att=att->next)
+        tree->setProp((const char *)att->name, (const char *)xmlGetProp(node, att->name));
+
+    for (xmlNodePtr child=xmlFirstElementChild(node); child!=nullptr; child=xmlNextElementSibling(child))
+        tree->addPropTree((const char *)child->name, createPtreeFromXmlNode(child));
+}
+
+IPropertyTree *createPtreeFromXmlNode(xmlNodePtr node)
+{
+    if (!node)
+        return nullptr;
+    Owned<IPropertyTree> tree = createPTree((const char *)node->name);
+    copyXmlNode(tree, node);
+    return tree.getClear();
+}
+
+static const char *queryAttrValue(xmlNodePtr node, const char *name)
+{
+    if (!node)
+        return nullptr;
+    xmlAttrPtr att = xmlHasProp(node, (const xmlChar *)name);
+    if (!att)
+        return nullptr;
+    if (att->type != XML_ATTRIBUTE_NODE || att->children==nullptr || (att->children->type != XML_TEXT_NODE && att->children->type != XML_CDATA_SECTION_NODE))
+        return nullptr;
+    return (const char *) att->children->content;
```

Comment:
So an attribute's value is its child node if that node is TEXT or CDATA?

Reply:
It's of type XML_ATTRIBUTE_NODE content.  A CDATA node is a different type.  In this "queryAttrValue() function I am relying on the fact that these are simple attributes and not a compounding of multiple parts like some elements and perhaps more complex attributes might be.  If we try and use this function for uses that don't match this assumption we may need to make it less efficient and have to allocate and delete memory.  For now this one method errors on the side of efficiency.